### DISCUSSION TITLE
Node Editor

### DISCUSSION THREAD

**Discussion by XBagon** [emilk/egui#166] [General]
I wondered how feasible a node editor built on top of `egui` would be. I'm guessing it would need some internal work, to be able to draw connections and the connection points on windows?

**Comment by emilk** [comment]
It should be fairly feasible, either by doing manual drawing and widget layout, or by using `Window`s as you nodes. I have experimented with the latter in the branch `demo-node-graph`: https://github.com/emilk/egui/blob/demo-node-graph/egui_demo_lib/src/apps/node_graph.rs but haven't had time to finish that up (lacks the adding and removing of edges).

**Reply by torsteingrindvik** [reply]
Hi, I found this thread randomly while checking if egui supports drawing graphs.
Seems it might be on its way.

Just chiming in to show interest.

Thanks for the great lib!

**Reply by samcarey** [reply]
I'm also interested in such a feature. It would be nice if the graph could be zoomable with the mouse wheel, but I'm not sure how difficult that would be.

**Reply by francesco-cattoglio** [reply]
I am super interested in something like this, because right now my software relies on imgui to have a functioning node graph editor and I am trying to make my project rust-only.
I would love to get involved in developing such a node editor in egui (having stuff like a scrollable area, zoomable graph, and such). Would you like to invest some time helping me getting on board for this?

**Comment by emilk** [comment]
Has anyone tried https://github.com/haighcam/egui_nodes ?

**Reply by francesco-cattoglio** [reply]
I did, and there is one major point that I would like to address: there are many situations in which implementing correctly a node graph is as much work as implementing the egui window system from scratch: as an example, right now if have the node "1" containing a text edit widget and put a node "2" in the front, you can click through node "2" which is in front node and start writing in the text field, even node "1" is completely hidden. Similarly, hovering and similar interactions also require you to re-write lots of code to mimic the behaviour of windows. I am using imnodes myself right now and I appreciate the effort of somebody trying to port it over to egui, but I feel like a different approach might yield a better result.

**Reply by emilk** [reply]
Yeah, being able to reuse `egui::Window` or `egui::Area` would be great (like I started doing in https://github.com/emilk/egui/blob/demo-node-graph/egui_demo_lib/src/apps/node_graph.rs)

**Reply by francesco-cattoglio** [reply]
I started experimenting and found out that one can allow windows to move "out of frame bounds" by setting `drag_bounds(Rect::EVERYTHING)` at window build time, which is a very much needed feature. The only drawback is that in my little experiment it could go above the side panels, but I am pretty sure it is something that can be fixed rather easily.

I think this could be a good way to approach the implementation. I will spend some time exploring and then report back when I have a little more knowledge about the whole thing.

**Comment by ricardoalcantara** [comment]
Let me revive this thread! I checked this repo https://github.com/haighcam/egui_nodes and it's broken, last update uses egui 0.16. I liked your implementation https://github.com/emilk/egui/blob/demo-node-graph/egui_demo_lib/src/apps/node_graph.rs, it was the simplest and the one I could easily understand, and also had tried something like that before too, just a few things I would like to point, the Windows cannot go out side the canvas and it limits the size of the graph, with that it would be nice to have a way to scroll it. Do you think it's feasible to achieve that?

**Comment by wjbuys** [comment]
Check out https://github.com/setzer22/egui_node_graph

**Reply by milewski** [reply]
It's dead now :(

**Reply by theoparis** [reply]
I forked egui_nodes by haighcam here if you are interested: https://github.com/tinted-software/egui_nodes

**Reply by trevyn** [reply]
Also available and maintained: https://github.com/trevyn/egui_node_graph2
